File: Solutions/1577. Number of Ways Where Square of Number Is Equal to Product of Two Numbers/1577. Number of Ways Where Square of Number Is Equal to Product of Two Numbers.cpp
class Solution {
public:
    int numTriplets(vector<int>& nums1, vector<int>& nums2) {
        return foo(nums1,nums2)+foo(nums2,nums1);
    }
    int foo(vector<int>& nums1,vector<int>& nums2){
        unordered_map<int,int> m;
        int res=0;
        for(auto& n2:nums2)
            ++m[n2];
        for(auto& n1:nums1){
            long target=pow(n1,2);
            for(auto& [k,cnt]:m){
                if(target%k or m.count(target/k)==0)
                    continue;
                if(target/k==k)
                    res+=cnt*(cnt-1);
                else
                    res+=cnt*m[target/k];
            }
        }
        return res/2;
    }
};

class Solution {
public:
        int numTriplets(vector<int>& n1, vector<int>& n2) {
        return accumulate(begin(n1), end(n1), 0, [&](int s, long n) { return s + twoProduct(n * n, n2); }) 
            + accumulate(begin(n2), end(n2), 0, [&](int s, long n) { return s +  twoProduct(n * n, n1); });
    }
    int twoProduct(long i, vector<int> &n) {
        unordered_map<int, int> m;
        int cnt = 0;
        for (auto v : n) {
            if (i % v == 0)
                cnt += m[i / v];
            ++m[v];
        }
        return cnt;
    } 
};

File: Solutions/1577. Number of Ways Where Square of Number Is Equal to Product of Two Numbers/1577. Number of Ways Where Square of Number Is Equal to Product of Two Numbers.java
class Solution {
    public int numTriplets(int[] nums1, int[] nums2) {
        return (int)(count(nums1,nums2)+count(nums2,nums1));
    }
    private long count(int[] nums1,int[] nums2){
        long res=0;
        HashMap<Long,Long> map=new HashMap<>();
        for(long n2:nums2){
            long tmp=map.getOrDefault(n2,0l);
            ++tmp;
            map.put(n2,tmp);
        }
        for(long n1:nums1){
            long target=n1*n1;
            for(var entry:map.entrySet()){
                long key=entry.getKey();
                long value=entry.getValue();
                if(target%key!=0||!map.containsKey(target/key))
                    continue;
                if(target/key==key)
                    res+=value*(value-1);
                else
                    res+=value*map.get(target/key);
            }
        }
        return res/2;
    }
}


class Solution {
    public int numTriplets(int[] n1, int[] n2) {
        long res = 0;
        for (long n : n1)
            res += twoProduct(n * n, n2);
        for (long n : n2)
            res += twoProduct(n * n, n1);
        return (int)res;
    }
    long twoProduct(long i, int[] n) {
        Map<Long, Long> m = new HashMap<>();
        long cnt = 0;
        for (long v : n) {
            if (i % v == 0)
                cnt += m.getOrDefault(i / v, 0l);
            m.put(v, m.getOrDefault(v, 0l) + 1);
        }
        return cnt;
    }  
}